### PR TITLE
Form validation and error handling

### PR DESCRIPTION
**Pull request by tanadendere**
- Implemented form validation to ensure data integrity and prevent submission of invalid input for login, register as well as adding trips and itinerary items.
- Added state management to check if a user's email already exists before allowing registration 
- Added state management for user existence validation during login to ensure only registered users can access the system.
- Also some code refactoring to match updates from main branch

### CODE REVIEW

File: src/app/components/home/add-trip/add-trip.component.html [hunk 1]
```diff
@@ -1,25 +1,40 @@
...
+    }
   </div>
+  <!-- in UI feature add *required instead -->
   <div>
     <!-- TO-DO: in ui/css toggle the form display -->
     <button>Cancel</button>
   </div>
   <div>
-    <button type="submit">Create trip</button>
+    <button type="submit" [disabled]="!form.valid">Create trip</button>
```
Comment (Musa-Mabasa) [inline src/app/components/home/add-trip/add-trip.component.html]: You can use `form.invalid` here

File: src/app/userManagement/components/register/register.component.ts [hunk 2]
```diff
@@ -1,54 +1,137 @@
...
+  confirmPasswordValidator(
+    password: string,
+    matchingPassword: string
+  ): ValidatorFn {
+    return (abstractControl: AbstractControl) => {
+      const passwordControl = abstractControl.get(password);
+      const matchingPaaswordControl = abstractControl.get(matchingPassword);
+
+      if (
+        matchingPaaswordControl!.errors &&
```
Comment (Musa-Mabasa) [inline src/app/userManagement/components/register/register.component.ts]: Avoid assertions here. Do an if check on `matchingPasswordControl`

File: src/app/userManagement/components/register/register.component.ts [hunk 3]
```diff
@@ -1,54 +1,137 @@
...
+        !matchingPaaswordControl!.errors?.['confirmedValidator']
+      ) {
+        return null;
+      }
+
+      if (passwordControl!.value !== matchingPaaswordControl!.value) {
```
Comment (Musa-Mabasa) [inline src/app/userManagement/components/register/register.component.ts]: Same thing here with the assertion

**Review by Musa-Mabasa — CHANGES_REQUESTED** [review]
I left some suggestion comments. I request a change with the assertions. Either than that, everything else is good 👍

File: src/app/userManagement/components/register/register.component.ts [hunk 4]
```diff
@@ -95,21 +95,23 @@ export class RegisterComponent {
...
-      }
-
-      if (passwordControl!.value !== matchingPaaswordControl!.value) {
-        const error = { confirmedValidator: 'Passwords do not match.' };
-        matchingPaaswordControl!.setErrors(error);
-        return error;
-      } else {
-        matchingPaaswordControl!.setErrors(null);
-        return null;
+      if (matchingPaaswordControl && passwordControl) {
```
Comment (Musa-Mabasa) [inline src/app/userManagement/components/register/register.component.ts:98]: 🫡

**Review by Musa-Mabasa — APPROVED** [review]
Everything looks good 👍

File: src/app/userManagement/components/login/login.component.ts [hunk 5]
```diff
@@ -21,22 +22,48 @@ export class LoginComponent {
...
+        Validators.pattern(/^[a-zA-Z0-9._-]+@[a-zA-Z0-9.-]+.[a-zA-Z]{2,4}$/),
+      ],
+    ],
+    password: [
+      '',
+      [
+        Validators.required,
+        Validators.pattern(/^(?=.*[A-Za-z])(?=.*[0-9])(?=.*[@$!%*#?&]).{3,}$/),
+      ],
+    ],
```
Comment (yanevdw) [inline src/app/userManagement/components/login/login.component.ts:42]: Nice!

File: src/app/userManagement/components/register/register.component.html [hunk 6]
```diff
@@ -1,31 +1,102 @@
...
+      <div>Name is required</div>
+      } @if (name?.hasError('maxlength')) {
+      <div>
+        Names can't be longer than 30 characters. Please shorten your name and
+        try again.
+      </div>
+      } @if (name?.hasError('pattern')) {
+      <div>
+        Please enter a valid name. Only letters, spaces, and apostrophes are
+        allowed.
```
Comment (yanevdw) [inline src/app/userManagement/components/register/register.component.html:22]: Great error handling here!

**Review by yanevdw — APPROVED** [review]
Amazing job Tana! Your validation and error handling is exceptional.

**Pull request merged**